### PR TITLE
feat(api): inbound dq from cw collects doc refs from metadata files

### PR DESCRIPTION
**Pull request by RamilGaripov**
refs. metriport/metriport-internal#2240

### Description
- We no longer query the FHIR server for CW shareback
- Query S3 for metadata files on the `uploads` folder, parse them, and create FHIR DocumentReferences, which we then return to CW

### Testing

- Staging
  - [x] Create 2 different orgs on staging CW. Insert the same patient for both. Add document from one org, pull from the other. 
    - [x] Test for structured doc uploads (CDA/CCD)
    - [x] Test for unstructured doc uploads (images, etc)

### Release Plan

- [ ] Merge this

### CODE REVIEW

File: packages/api/src/external/commonwell/proxy/cw-process-request.ts [hunk 1]
```diff
@@ -45,13 +51,14 @@ export async function processRequest(req: Request): Promise<Bundle<Resource>> {
...
   const { cxId, patientId } = await getPatientAndCxFromRequest(req);
-
   const { resource, count, params } = fromHttpRequestToFHIR(req);
 
   log(
     `UPDATED resource: ${resource} / cx ${cxId} / patient ${patientId} ` +
       `/ count : ${count}, params: ${params.toString()}`
   );
+
+  // TODO: Remove when done comparing
```
Comment (leite08) [inline packages/api/src/external/commonwell/proxy/cw-process-request.ts:61]: left-over?

File: packages/api/src/external/commonwell/proxy/cw-process-request.ts [hunk 2]
```diff
@@ -45,13 +51,14 @@ export async function processRequest(req: Request): Promise<Bundle<Resource>> {
...
   const rawResources = await queryFHIRServer({
```
Comment (leite08) [inline packages/api/src/external/commonwell/proxy/cw-process-request.ts:62]: Missed removing this and the oldBundle below?

File: packages/core/src/shareback/metadata/constants.ts [hunk 3]
```diff
@@ -0,0 +1,29 @@
+/*
+CONFIDENTIALITY_CODE is always N, or normal, indicating its normal PHI
+FORMAT_CODE is always urn:ihe:iti:xds:2017:mimeTypeSufficient, which is sufficient for all document types.
```
Comment (leite08) [inline packages/core/src/shareback/metadata/constants.ts:3]: Can we associate these comments to the respective field?

File: packages/api/src/external/commonwell/proxy/cw-process-request.ts (hunk 2 above)
Comment (RamilGaripov) [reply packages/api/src/external/commonwell/proxy/cw-process-request.ts:62]: Huh.. I must've forgot to push my last changes where I cleaned it all up 😬

File: packages/api/src/external/commonwell/proxy/cw-process-request.ts [hunk 4]
```diff
@@ -38,30 +41,41 @@ const allowedQueryParams = ["status", "date", "_include", "_summary"];
...
 
   const { cxId, patientId } = await getPatientAndCxFromRequest(req);
-
   const { resource, count, params } = fromHttpRequestToFHIR(req);
 
+  // TODO: We might need to introduce filtering based on the provided query params.
   log(
     `UPDATED resource: ${resource} / cx ${cxId} / patient ${patientId} ` +
       `/ count : ${count}, params: ${params.toString()}`
   );
```
Comment (RamilGaripov) [inline packages/api/src/external/commonwell/proxy/cw-process-request.ts:58]: These params are now only used in this log, so we'll always be returning everything we have that our CX uploaded. Idk how important this is right now to introduce filtering based on these params to the new flow. The only one I'm seeing that makes sense to me is the `date` filter.

File: packages/api/src/external/commonwell/proxy/cw-process-request.ts [hunk 5]
```diff
@@ -38,30 +41,41 @@ const allowedQueryParams = ["status", "date", "_include", "_summary"];
...
+ * - return a bundle with the DocumentReferences
  */
 export async function processRequest(req: Request): Promise<Bundle<Resource>> {
   log(`ORIGINAL URL: ${req.url}`);
 
   const { cxId, patientId } = await getPatientAndCxFromRequest(req);
-
   const { resource, count, params } = fromHttpRequestToFHIR(req);
 
+  // TODO: We might need to introduce filtering based on the provided query params.
```
Comment (leite08) [inline packages/api/src/external/commonwell/proxy/cw-process-request.ts:54]: This is a good point. The previous version was applying most filters to the call to the FHIR server.

Why can't we also apply those filters here?

File: packages/core/src/external/carequality/shared.ts
Comment (RamilGaripov) [inline packages/core/src/external/carequality/shared.ts:1]: Just moving some of the constants from this file to `packages/core/src/shareback/metadata/constants.ts`

File: packages/core/src/shareback/metadata/constants.ts
Comment (RamilGaripov) [inline packages/core/src/shareback/metadata/constants.ts:1]: No new data here. Everything came from `packages/core/src/external/carequality/shared.ts`

File: packages/core/src/shareback/metadata/constants.ts (hunk 3 above)
Comment (RamilGaripov) [reply packages/core/src/shareback/metadata/constants.ts:3]: will do

File: packages/api/src/external/commonwell/proxy/cw-process-request.ts (hunk 4 above)
Comment (leite08) [reply packages/api/src/external/commonwell/proxy/cw-process-request.ts:58]: This is a good point. The previous version was applying most filters to the call to the FHIR server.

Why can't we also apply those filters here?

File: packages/api/src/external/commonwell/proxy/cw-process-request.ts [hunk 6]
```diff
@@ -187,3 +180,60 @@ function replaceAttachmentURL(url: string): string {
...
+          case "gt":
+            return docDate.isAfter(filterDate);
+          case "ge":
+            return docDate.isSame(filterDate) || docDate.isAfter(filterDate);
+          case "le":
+            return docDate.isSame(filterDate) || docDate.isBefore(filterDate);
+          case "sa":
+            return docDate.isAfter(filterDate);
+          case "eb":
+            return docDate.isBefore(filterDate);
```
Comment (leite08) [inline packages/api/src/external/commonwell/proxy/cw-process-request.ts:219]: Aren't those:
- `sa`: same or after
- `eb`: equal or before

...in which case we'd need to add the check for sameness, right?

File: packages/api/src/external/commonwell/proxy/cw-process-request.ts [hunk 7]
```diff
@@ -34,8 +32,9 @@ const countParam = "_count";
 /**
  * CW might send `category` but we've seen it as `('34133-9%5E%5E2.16.840.1.113883.6.1')` which is not valid
  * for the FHIR server, and there's nothing on CW's spec about it, so we're not going to support it for now.
+ * CW might also send `_include`, but since we're not using the FHIR server, we will always return DocumentReferences with the subject (Patient) and the contained Organization
  */
-const allowedQueryParams = ["status", "date", "_include", "_summary"];
+const allowedQueryParams = ["status", "date", "_summary"];
```
Comment (leite08) [inline packages/api/src/external/commonwell/proxy/cw-process-request.ts:37]: Only now I noticed something: if they don't add any of these, we'll throw 400, and unless there's something that requires those params on the request, I think we shouldn't.

Thoughts?

Comment (RamilGaripov) [reply packages/api/src/external/commonwell/proxy/cw-process-request.ts:37]: Ya, the only reason for this that I'm seeing is that the `patient.identifier` ends up in the query parameters. But it's already being parsed in the first step of the request processing. So _I think_ it's safe to remove this constraint

**Review by leite08 — APPROVED** [review]
:shipit:

**Pull request merged**